File: engine/src/main/java/com/terralite/engine/terrain/BlockStorage.java
package com.terralite.engine.terrain;

import java.util.Collection;
import java.util.OptionalInt;

public interface BlockStorage {
    BlockState get(BlockPos pos);

    void set(BlockPos pos, BlockState state);

    BlockState remove(BlockPos pos);

    default OptionalInt stateId(BlockPos pos) {
        return OptionalInt.empty();
    }

    boolean contains(BlockPos pos);

    Collection<BlockPos> positions();

    int size();
}

File: engine/src/commonMain/kotlin/com/terralite/engine/terrain/BlockStorage.kt
package com.terralite.engine.terrain

interface BlockStorage {
    fun get(pos: BlockPos): BlockState
    fun set(pos: BlockPos, state: BlockState)
    fun remove(pos: BlockPos): BlockState
    fun stateId(pos: BlockPos): Int? = null
    fun contains(pos: BlockPos): Boolean
    fun positions(): Collection<BlockPos>
    fun size(): Int
}
